File: clients/broker_preflight.py
"""Pure, bearer-safe broker preflight helpers."""

from dataclasses import dataclass
import re
from urllib.parse import urlparse


class BrokerPreflightError(ValueError):
    """Raised when a non-secret broker setting is absent or invalid."""


_API_SCOPE_PATTERN = re.compile(r"^api://[^/\s]+/.+$")


@dataclass(frozen=True)
class BrokerPreflightConfig:
    base: str
    scope: str


def _required(environ, name):
    value = environ.get(name)
    if not isinstance(value, str) or not value.strip():
        raise BrokerPreflightError(f"{name} is required")
    return value.strip()


def validate_broker_config(environ):
    """Return trimmed required broker settings without introducing defaults."""
    base = _required(environ, "BROKER_BASE")
    scope = _required(environ, "BROKER_SCOPE")
    parsed = urlparse(base)
    if parsed.scheme != "https" or not parsed.netloc:
        raise BrokerPreflightError("BROKER_BASE must be an HTTPS URL")
    if (scope.startswith("api://") and not _API_SCOPE_PATTERN.fullmatch(scope)) or (
        not scope.startswith("api://") and any(character.isspace() for character in scope)
    ):
        raise BrokerPreflightError("BROKER_SCOPE is malformed")
    return BrokerPreflightConfig(base=base.rstrip("/"), scope=scope)


def run_preflight(config, acquire_token, invoke_preflight):
    """Invoke preflight with injected transport and expose only safe response fields."""
    token = acquire_token()
    response = invoke_preflight(config.base, token)
    return (response["status"], response["correlationId"])

File: clients/node/broker-preflight.mjs
const apiScopePattern = /^api:\/\/[^/\s]+\/.+$/;

function required(env, name) {
  const value = env[name];
  if (typeof value !== 'string' || value.trim() === '') {
    throw new Error(`${name} is required`);
  }
  return value.trim();
}

/** Load the non-secret broker settings used by the preflight adapter. */
export function loadBrokerConfig(env = process.env) {
  const rawBase = required(env, 'BROKER_BASE');
  const scope = required(env, 'BROKER_SCOPE');

  let parsed;
  try {
    parsed = new URL(rawBase);
  } catch {
    throw new Error('BROKER_BASE must be an HTTPS URL');
  }
  if (parsed.protocol !== 'https:' || !parsed.hostname || parsed.username || parsed.password || parsed.search || parsed.hash) {
    throw new Error('BROKER_BASE must be an HTTPS URL');
  }
  if ((scope.startsWith('api://') && !apiScopePattern.test(scope)) || (!scope.startsWith('api://') && /\s/.test(scope))) {
    throw new Error('BROKER_SCOPE is malformed');
  }

  return { base: rawBase.replace(/\/+$/, ''), scope };
}

/**
 * Acquire a token and invoke an authenticated preflight through injected seams.
 * Deliberately return only the safe response fields.
 */
export async function runPreflight({ config, acquireToken, invokePreflight }) {
  const token = await acquireToken();
  const response = await invokePreflight(config.base, token);
  return { status: response.status, correlationId: response.correlationId };
}
